### PR TITLE
ledger-sim: support all-reference-script setup

### PR DESCRIPTION
**Pull request by chfanghr**

### CODE REVIEW

File: ledger-sim/src/Ledger/Sim/Types/LedgerConfig.hs [hunk 1]
```diff
@@ -12,10 +13,13 @@ import PlutusLedgerApi.V2 qualified as PlutusV2
 import Ledger.Sim.Types.CostModel (PlutusCostModel (PlutusCostModel), mkEvaluationContext)
 import PlutusLedgerApi.Common (ExBudget)
 
+data ScriptMode = ScriptMode'AllowWitness | ScriptMode'MustBeReference
```
Comment (szg251) [inline ledger-sim/src/Ledger/Sim/Types/LedgerConfig.hs:16]: why not simply `Witness` and `Reference`, and verify that the script is one or the other?

File: ledger-sim/src/Ledger/Sim/Types/LedgerConfig.hs [hunk 2]
```diff
@@ -12,10 +13,13 @@ import PlutusLedgerApi.V2 qualified as PlutusV2
...
+
 data LedgerConfig ctx = LedgerConfig
   { lc'evaluationContext :: Plutus.EvaluationContext
   , lc'scriptStorage :: !(Map PlutusV2.ScriptHash PlutusV2.ScriptForEvaluation)
   , lc'maxExBudget :: Maybe ExBudget
+  , lc'scriptMode :: ScriptMode
```
Comment (szg251) [inline ledger-sim/src/Ledger/Sim/Types/LedgerConfig.hs:22]: I think this should be an option for each script, and not global.

File: ledger-sim/src/Ledger/Sim/Types/LedgerConfig.hs (hunk 1 above)
Comment (chfanghr) [reply ledger-sim/src/Ledger/Sim/Types/LedgerConfig.hs:16]: Cause some people simply don't care about how their scripts are presented. Unbox is special in a sense that it relies on the reference input order to function properly. So basically, `AllowWitness` actually means "unchecked" here.

File: ledger-sim/src/Ledger/Sim/Types/LedgerConfig.hs (hunk 2 above)
Comment (chfanghr) [reply ledger-sim/src/Ledger/Sim/Types/LedgerConfig.hs]: Implemented!

File: ledger-sim/src/Ledger/Sim/Validation/Stateful.hs [hunk 3]
```diff
@@ -96,13 +103,21 @@ validateValidRange state =
 
 --------------------------------------------------------------------------------
 
-newtype InvalidRedeemersError = InvalidRedeemers'ScriptRequiredForEvaluationNotAvailable ScriptHash
+data RedeemerKind = RedeemerKind'Spending | RedeemerKind'Minting
   deriving stock (Show, Eq)
 
-validateRedeemers :: LedgerConfig ctx -> Validator InvalidRedeemersError ([TxInInfo], Value)
+data InvalidRedeemerErrorKind
```
Comment (bladyjoker) [inline ledger-sim/src/Ledger/Sim/Validation/Stateful.hs:109]: Why are you saying `Kind` here? Why not `InvalidRedeemerError`? Kind/Type/Sort is smtn redundant as it's captured by the fact that it's a sum type

File: ledger-sim/src/Ledger/Sim/Validation/Stateful.hs [hunk 4]
```diff
@@ -117,13 +132,45 @@ validateRedeemers config =
...
+            (validateWith $ validateFail . InvalidRedeemersError k InvalidRedeemerErrorKind'UnknownScript)
+            (validateScriptHash k)
+
+    validateScriptHash :: RedeemerKind -> Validator InvalidRedeemersError (S.Set ScriptHash, ScriptMode, ScriptHash)
+    validateScriptHash k = validateWith $ \(availableReferenceScripts, scriptMode, scriptHash) ->
+      contramap (const scriptHash) $
+        validateIf
+          ( case scriptMode of
+              ScriptMode'Unchecked -> const True
+              ScriptMode'MustBeReference -> flip S.member availableReferenceScripts
```
Comment (bladyjoker) [inline ledger-sim/src/Ledger/Sim/Validation/Stateful.hs:168]: Yeah, AsWitness as AsReference makes sense here I think

**Review by bladyjoker — APPROVED** [review]
Another naming review :D

File: ledger-sim/src/Ledger/Sim/Submission.hs [hunk 5]
```diff
@@ -142,7 +142,7 @@ mustFindScriptByHash h = do
 
   case M.lookup h scriptStorage of
     Nothing -> shouldHaveBeenCaughtByValidation "script with the given hash not found in script storage"
```
[the remark this answers is not in the document]
Comment (chfanghr) [reply ledger-sim/src/Ledger/Sim/Submission.hs:144]: `shouldHaveBeenCaughtByValidation` simply aborts

**Pull request merged**